tsl/src/gapfill/exec.c

```diff
@@ -244,6 +291,178 @@ is_simple_expr(Expr *node)
+	FuncExpr   *time_bucket = copyObject(linitial(cscan->custom_private));
+	Datum		value;
+	bool		isnull;
+
+	if (!is_simple_expr(expr))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid time_bucket_gapfill argument: start must be a simple expression")));
+
+	time_bucket->args = list_make2(linitial(time_bucket->args), expr);
+	value = gapfill_exec_expr(state, (Expr *) time_bucket, &isnull);
+
+	/* start expression must not evaluate to NULL */
+	if (isnull)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid time_bucket_gapfill argument: start cannot be NULL"),
+				 errhint("You can either pass start and finish as arguments or in the WHERE clause")));
+
+	return gapfill_datum_get_internal(value, state->gapfill_typid);
+}
+
+static int64
+get_boundary_expr_value(GapFillState *state, GapFillBoundary boundary, Expr *expr)
+{
+	Datum		arg_value;
+	bool		isnull;
+
+	/*
+	 * add an explicit cast here if types do not match
+	 */
+	if (exprType((Node *) expr) != state->gapfill_typid)
+	{
+		Oid			cast_oid = get_cast_func(exprType((Node *) expr), state->gapfill_typid);
+
+		expr = (Expr *) makeFuncExpr(cast_oid, state->gapfill_typid, list_make1(expr), InvalidOid, InvalidOid, 0);
+	}
+
+	arg_value = gapfill_exec_expr(state, expr, &isnull);
+
+	if (isnull)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid time_bucket_gapfill argument: %s cannot be NULL", boundary == GAPFILL_START ? "start" : "finish"),
+				 errhint("You can either pass start and finish as arguments or in the WHERE clause")));
+
+	return gapfill_datum_get_internal(arg_value, state->gapfill_typid);
+}
+
+static int64
+infer_gapfill_boundary(GapFillState *state, GapFillBoundary boundary)
+{
+	CustomScan *cscan = castNode(CustomScan, state->csstate.ss.ps.plan);
+	FuncExpr   *func = linitial(cscan->custom_private);
+	FromExpr   *jt = lfourth(cscan->custom_private);
+	List	   *vars;
+	ListCell   *lc;
+	TypeCacheEntry *tce = lookup_type_cache(state->gapfill_typid, TYPECACHE_BTREE_OPFAMILY);
+	int			strategy;
+	Oid			lefttype,
+				righttype;
+
+	int64		boundary_value = 0;
+	bool		boundary_found = false;
+
+	/*
+	 * if the second argument to time_bucket_gapfill is not a column reference
+	 * we cannot match WHERE clause to the time column, we don't check for Var
+	 * directly though cause it might be wrapped in a cast
+	 */
+	vars = pull_var_clause(lsecond(func->args), 0);
+	if (list_length(vars) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid time_bucket_gapfill argument: ts needs to refer to column if no start or finish is supplied"),
+				 errhint("You can either pass start and finish as arguments or in the WHERE clause")));
+
+	foreach(lc, (List *) jt->quals)
+	{
+		OpExpr	   *opexpr = lfirst(lc);
+		Var		   *var;
+		Expr	   *expr;
+		Oid			op;
+		int64		value;
+
+		if (!IsA(opexpr, OpExpr))
```

Comment:
This should really be before the cast to `OpExpr  *`. With how it is now, the compiler is allowed to start looking at the `OpExpr` before checking if `IsA(opexpr, OpExpr)`, which could segfault if it isn't.

Reply:
changed